### PR TITLE
Add support for Sun Grid Engine batch system

### PR DESCRIPTION
**Pull request by trygveasp**
Based on the template for PBS but needs some modifications.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/troika?pullRequest=9) <br/>All committers have signed the CLA.

File: src/troika/sites/base.py
Comment (oiffrig) [inline src/troika/sites/base.py]: I don't think this belongs here. My suggestion would be to implement it as a hook (cf. `src/troika/hooks/common.py`), adding a `remove_previous_output` method to the `Site` class, that the hook can call.

Is this a requirement for SGE? If so, we can also call the method directly from `SGESite.submit`

File: src/troika/sites/sge.py
Comment (oiffrig) [inline src/troika/sites/sge.py]: From what I see in the docs, the option corresponding to wall time limit is `s_rt` (or `h_rt` for a hard kill straight away)

```suggestion
        "walltime": b"-l s_rt=%s",
```

File: src/troika/controllers/base.py
Comment (oiffrig) [inline src/troika/controllers/base.py]: This may have unintended consequences. I see the `-j y` to join output and error streams, but only in some docs and not all, not sure whether it is always available. This would make it similar to PBS

File: src/troika/sites/sge.py
Comment (oiffrig) [inline src/troika/sites/sge.py]: I don't think this command exists for SGE, unfortunately

File: src/troika/sites/sge.py
Comment (oiffrig) [inline src/troika/sites/sge.py]: No need for a loop in this case, since a kill sequence is not supported

File: src/troika/sites/sge.py
Comment (oiffrig) [inline src/troika/sites/sge.py]:
```suggestion
        self._connection.execute(self._qstat + ["-j", jid], stdout=outf, dryrun=dryrun)
```

**Review by oiffrig — CHANGES_REQUESTED** [review]
Thanks for the PR! I don't have access to a Grid Engine cluster so I can only judge based on the documentation I've found (and there seem to be different flavours), happy to merge once we have ironed out the few details

File: src/troika/controllers/base.py
Comment (trygveasp) [reply src/troika/controllers/base.py]: Sadly I have not been able to get the -y option to work properly on our SGE system. What do you mean by similar to PBS? Do you have an example, use case or suggestion on what to do?

File: src/troika/sites/sge.py
Comment (trygveasp) [reply src/troika/sites/sge.py]: I modify it to h_rt. It's what we normally use and is a bit safer

File: src/troika/sites/base.py
Comment (trygveasp) [reply src/troika/sites/base.py]: I have added a hook and implemented it in troika.sites.base.Site as it could be used by other sites
It is not a hard requirement so it doesn't need to be in submit for SGE. However, logs are default appended and it makes them confusing.

File: src/troika/sites/sge.py
Comment (trygveasp) [reply src/troika/sites/sge.py]: You are right. That was my conclusion as well, but with qsub -notify and trapping on SIGUSR1 you can at least trap the killing. I added -notify this in my site config, but could of course be a part of this Site class.

File: src/troika/sites/sge.py
Comment (trygveasp) [reply src/troika/sites/sge.py]: Yep, I was lazy and made it an empty loop. Should I return "KILLED" as abort status from the method?

File: src/troika/sites/sge.py
Comment (trygveasp) [reply src/troika/sites/sge.py]: I had this in my site config. But I have modified it to your suggestion.

**Comment by trygveasp** [comment]
I am working on a modified version of the PR but I have a couple of questions and would also like to test it first before I update the PR.

File: src/troika/controllers/base.py
Comment (trygveasp) [reply src/troika/controllers/base.py]: I have removed it now.

File: src/troika/sites/sge.py
Comment (oiffrig) [reply src/troika/sites/sge.py]: Looks good to me now :)

File: src/troika/sites/sge.py
Comment (oiffrig) [reply src/troika/sites/sge.py]: Do we want to add this by default? Another option would be to add a directive, e.g. `"notify_end": b"-notify"` that the user may set if they want to get a signal. I don't mind too much since it is pretty harmless.

File: src/troika/controllers/base.py
Comment (oiffrig) [reply src/troika/controllers/base.py]: I was hoping to make the behaviour of writing to the output file only could work the same way as in the PBS site where we use "-j oe". Is the current version working for you?

**Review by oiffrig — APPROVED** [review]
Thanks!

Happy to merge now, but please do tell me if your tests are not successful. I'll wait for your green light, in case you want to add more changes.

File: src/troika/controllers/base.py
Comment (trygveasp) [reply src/troika/controllers/base.py]: It works by adding -j y either in the job header or by using the translator and add:
translators: ["join_output_error"]
in the site config. So I can stay with this and I don't need the change in the base controller

File: src/troika/sites/sge.py
Comment (trygveasp) [reply src/troika/sites/sge.py]: I removed -notify to the qsub command for sge. However, I don't get it to work directly as a directive unless I specify the directive as:
`"notify_end": b"-notify %s",`
and in the job specify:
`#TROIKA notify_end=`

This is because the parsing expects key,value pairs and I have to handle an empty value. I could maybe add an extra translator or a pre-submit hook or something similar. What do you think?

Comment (trygveasp) [reply src/troika/sites/sge.py]: It works also by specifying `qsub_command: ["qsub", "-notify"] ` to the site config. My site now looks like this:

```
  ppi-rhel8:
    type: sge
    connection: local
    pre_submit: ["create_output_dir", "remove_previous_output"]
    preprocess: ["remove_top_blank_lines"]
    translators: ["join_output_error"]
    qsub_command: ["qsub", "-notify"]
    qdel_command: 'qdel'
```
Maybe this is ok and I should just remove notify from the SGE directives....

**Comment by trygveasp** [comment]
It works as it should now, but I am not able to trap killed jobs in ecflow when they are in submitted state. The jobs are killed as they should, but this is not signaled to ecflow. I will have a look to see if I can fix this, but I would acknowledge information on how you do this!

**Comment by trygveasp** [comment]
> It works as it should now, but I am not able to trap killed jobs in ecflow when they are in submitted state. The jobs are killed as they should, but this is not signaled to ecflow. I will have a look to see if I can fix this, but I would acknowledge information on how you do this!

OK, by adding to the job files:
```
# TROIKA ecflow_name=%ECF_NAME%
# TROIKA ecflow_pass=%ECF_PASS%
# TROIKA ecflow_host=%ECF_HOST%
# TROIKA ecflow_port=%ECF_PORT%
```

and to the site config:
`post_kill: ["abort_on_ecflow"]`

it works as intended!

I then think I am good to go from my side!

**Comment by ernst-bablick** [comment]
Hi there! 
I found this by accident.
If your changes make it work for SGE, then it will also work with Open Cluster Scheduler (also available on GitHub). We forked our WLM from Univa Grid Engine and the command line interface is the same. If you are missing some functionality compared to PBS or Slurm, or need extensions to make your implementation easier, let us know. We can implement it in our batch scheduler...

**Review by oiffrig — APPROVED** [review]
Thanks @trygveasp!

**Pull request merged**
